### DISCUSSION TITLE
How to use pytorch framework in MLEngineTrainingOperator?

### DISCUSSION THREAD

**Discussion by MarsSu0618** [apache/airflow#12610] [Q&A] [answered]
## Problems
Hi, everyone. I have a question about Google Cloud ML-engine and airflow MLEngineTrainingOperator.
default parameters of MLEngineTrainingOperator have `runtime_version`, it can be specific version gcp ai platform runtime version. Default have tensorflow, sklearn and xgboost. But I want to use pytorch framework, i have to  self build pytorch container, and how to integrate airflow operator and training?

## Reference
* [Get Started with pytorch on Google cloud ML Engine](https://cloud.google.com/ai-platform/training/docs/getting-started-pytorch)
* [Google cloud runtime version](https://cloud.google.com/ai-platform/training/docs/runtime-version-list)

Hope someone can help. I will thanks a lot.

**Comment by turbaszek** [comment]
I'm not an expert but. The new `MLEngineStartTrainingJobOperator` from providers packages have `master_config` parameter which 
according to [docs](https://cloud.google.com/ai-platform/training/docs/reference/rest/v1/projects.jobs#TrainingInput)
> Set masterConfig.imageUri only if you build a custom image. Only one of masterConfig.imageUri and runtimeVersion should be set. Learn more about configuring custom containers.

this config should be [ReplicaConfig](https://cloud.google.com/ai-platform/training/docs/reference/rest/v1/projects.jobs#ReplicaConfig) where you can specify `imageUri` - The Docker image to run on the replica. 



That being said I would try:
```python
master_config={"imageUri" : "gcr.io/my-pytorch-image-uri"}
```

**Reply by MarsSu0618** [reply]
@turbaszek 
Thanks for your help.
If my airflow version is 1.10.0. Can i specific paramette for `training_args` of `MLEngineTrainingOperator`?

For example:
```python
ml_training = MLEngineTrainingOperator(
    task_id='create_ml_job',
    job_id='ml_job',
    training_args=[
        '--master-image-uri', 'gcr.io/my-pytorch-image-uri'
    ]
......
)
```

Because gcloud ai-platform job submit command have this parameters. But i not sure whether is can work?
reference link : [gcloud ai-platform jobs submit training](https://cloud.google.com/sdk/gcloud/reference/ai-platform/jobs/submit/training)

**Reply by turbaszek** [reply]
@MarsSu0618 I'm afraid this won't work but you can try. Docs:
> Command-line arguments passed to the training application when it starts. If your job uses a custom container, then the arguments are passed to the container's ENTRYPOINT command.

I would recommend to use providers packages as the old operators won't be supported in future. 
http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow/latest/provider-packages.html#using-backport-providers-in-airflow-1-10

**Reply by MarsSu0618** [reply]
@turbaszek 
you are right. I can not specific `--master-image-uri` in `training_args`. 
the error message 
```
 "Field: runtime_version Error: Runtime version must be provided when the master Docker image URI is empty."
```
It seems have to specific runtime version. But gcp runtime version do not include pytorch framework. 
Then, We still use airflow and version 1.10 recently. No other solution, right?

**Reply by turbaszek** [reply]
Have you take a look at backport providers? 
https://airflow.apache.org/docs/stable/backport-providers.html

> Thanks to that, the operators from Airflow 2.0 can be used in Airflow 1.10 as separately installable packages with the constraint that those packages can only be used in Python 3.6+ environment.

If you can't install them, have you considered copying the code of operator from master branch an using it in your DAG?

**Reply by turbaszek** [reply]
Hey, @MarsSu0618 any update on this one?

**Reply by MarsSu0618** [reply]
@turbaszek 
I use other solution to solve my problem and it can be work. 
I have build customOperator to extent original MLEngineTrainingOperator with adding `master-image-uri`.

For example as follows:
```python

class MyOperator(BaseOperator):
    def __init__(self,
                 project_id,
                 job_id,
                 package_uris,
                 training_python_module,
                 training_args,
                 region,
                 scale_tier=None,
                 master_type=None,
                 master_image_uri=None, ## add master image uri flags
                 runtime_version=None,
                 python_version=None,
                 job_dir=None,
                 gcp_conn_id='google_cloud_default',
                 delegate_to=None,
                 mode='PRODUCTION',
                 *args,
                 **kwargs):
        super(MyOperator, self).__init__(*args, **kwargs)
        self._project_id = project_id
        self._job_id = job_id
        self._package_uris = package_uris
        self._training_python_module = training_python_module
        self._training_args = training_args
        self._region = region
        self._scale_tier = scale_tier
        self._master_type = master_type
        self._master_image_uri = master_image_uri ## add master image uri flags
        self._runtime_version = runtime_version
        self._python_version = python_version
        self._job_dir = job_dir
        self._gcp_conn_id = gcp_conn_id
        self._delegate_to = delegate_to
        self._mode = mode
........

        if self._runtime_version:
            training_request['trainingInput']['runtimeVersion'] = self._runtime_version
        
        ## add master_image_uri to TrainingInput 
        if self._master_image_uri:
            training_request['trainingInput']['masterConfig']['imageUri'] = self._master_image_uri

        if self._python_version:
            training_request['trainingInput']['pythonVersion'] = self._python_version

        if self._job_dir:
            training_request['trainingInput']['jobDir'] = self._job_dir

        if self._scale_tier is not None and self._scale_tier.upper() == "CUSTOM":
            training_request['trainingInput']['masterType'] = self._master_type
........
hook = MLEngineHook(
            gcp_conn_id=self._gcp_conn_id, delegate_to=self._delegate_to)

        # Helper method to check if the existing job's training input is the
        # same as the request we get here.
        def check_existing_job(existing_job):
            return existing_job.get('trainingInput', None) == \
                training_request['trainingInput']
```
We can reference original MLEngineTrainingOperator SourceCode and know how to modify and add `imageUri` for Pytorch.
Then, move this custom operator python script file to airflow plugings folder and import in your dag script file, it will be work.

**Reply by turbaszek** [reply]
Sounds good! Should we mark this question as closed?

**Reply by MarsSu0618** [reply]
@turbaszek  sure! Thanks a lot for your help.

**Reply by turbaszek** [reply]
@MarsSu0618 great, can you please close it / mark as answered? I cannot do this 😉

**Comment by MarsSu0618** [comment, marked as answer]
The answer as follows:

```python
class MyOperator(BaseOperator):
    def __init__(self,
                 project_id,
                 job_id,
                 package_uris,
                 training_python_module,
                 training_args,
                 region,
                 scale_tier=None,
                 master_type=None,
                 master_image_uri=None, ## add master image uri flags
                 runtime_version=None,
                 python_version=None,
                 job_dir=None,
                 gcp_conn_id='google_cloud_default',
                 delegate_to=None,
                 mode='PRODUCTION',
                 *args,
                 **kwargs):
        super(MyOperator, self).__init__(*args, **kwargs)
        self._project_id = project_id
        self._job_id = job_id
        self._package_uris = package_uris
        self._training_python_module = training_python_module
        self._training_args = training_args
        self._region = region
        self._scale_tier = scale_tier
        self._master_type = master_type
        self._master_image_uri = master_image_uri ## add master image uri flags
        self._runtime_version = runtime_version
        self._python_version = python_version
        self._job_dir = job_dir
        self._gcp_conn_id = gcp_conn_id
        self._delegate_to = delegate_to
        self._mode = mode
........

        if self._runtime_version:
            training_request['trainingInput']['runtimeVersion'] = self._runtime_version
        
        ## add master_image_uri to TrainingInput 
        if self._master_image_uri:
            training_request['trainingInput']['masterConfig']['imageUri'] = self._master_image_uri

        if self._python_version:
            training_request['trainingInput']['pythonVersion'] = self._python_version

        if self._job_dir:
            training_request['trainingInput']['jobDir'] = self._job_dir

        if self._scale_tier is not None and self._scale_tier.upper() == "CUSTOM":
            training_request['trainingInput']['masterType'] = self._master_type
........
hook = MLEngineHook(
            gcp_conn_id=self._gcp_conn_id, delegate_to=self._delegate_to)

        # Helper method to check if the existing job's training input is the
        # same as the request we get here.
        def check_existing_job(existing_job):
            return existing_job.get('trainingInput', None) == \
                training_request['trainingInput']
```
We can reference original MLEngineTrainingOperator SourceCode and know how to modify and add imageUri for Pytorch.
Then, move this custom operator python script file to airflow plugings folder and import in your dag script file, it will be work

**Comment by jahidhasanlinix** [comment]
@MarsSu0618 I was thinking of integrating my DL trained model into Spark airflow for DAGscheduling. Did you able to integrate it successfully. Can I have your source code to look into if its shareable?
